### PR TITLE
Refuse relative path for remote in coordinator

### PR DESCRIPTION
**Pull request by XxChang**
Only allow absolute path for remote daemon in coordinator,

modified from https://github.com/dora-rs/dora/pull/534

See option 2 in https://github.com/dora-rs/dora/issues/535

### CODE REVIEW

**Comment by XxChang** [comment]
@Gege-Wang please help me check it in remote daemon

**Comment by Gege-Wang** [comment]
@XxChang Hi, I think the machine who executes `dora start dataflow.yml` is the local machine.In this PR, the local machine is same as the coordinator, right?

**Comment by haixuanTao** [comment]
So this work on aws EC2 right?

And now you should be able to spawn multiple-daemon example without using the same folder structure right?

**Comment by phil-opp** [comment]
> I think the machine who executes `dora start dataflow.yml` is the local machine.In this PR, the local machine is same as the coordinator, right?

Good point!

@XxChang Could you add an additional check in `dora start` that the coordinator IP is local too? Then we can be sure that the coordinator and all relevant daemons run on the same machine.

**Comment by XxChang** [comment]
> > I think the machine who executes `dora start dataflow.yml` is the local machine.In this PR, the local machine is same as the coordinator, right?
> 
> Good point!
> 
> @XxChang Could you add an additional check in `dora start` that the coordinator IP is local too? Then we can be sure that the coordinator and all relevant daemons run on the same machine.

Sure, but if we do that. The "coordinator_addr" optional flag in `dora start` will be no sense

**Comment by phil-opp** [comment]
> > @XxChang Could you add an additional check in `dora start` that the coordinator IP is local too? Then we can be sure that the coordinator and all relevant daemons run on the same machine.
> 
> Sure, but if we do that. The "coordinator_addr" optional flag in `dora start` will be no sense

I meant as a requirement for absolute paths. Relative paths should be only allowed if the CLI, the coordinator, and the daemon(s) are all on the same machine.

So you can still use `--coordinator_addr`, but not in combination with relative paths.

**Comment by XxChang** [comment]
Add additional check in `dora start`. Only allow relative path to be used in the case that `dora-cli` and `dora-coordinator` are in the same machine.

Also modify the `ci.yml` to left enough space for `python-dataflow` test case.

**Review by haixuanTao — APPROVED** [review]
Looks good to me! 

Thanks! Looks very good!

Maybe i will let Philipp have a look and merge it

**Review by phil-opp — COMMENTED** [review]
Thanks for the updates and thanks for finding a way to free up enough space on CI!

I left one suggestion, otherwise this looks good to me.

File: libraries/core/src/descriptor/validate.rs [hunk 1]
```diff
@@ -12,18 +12,44 @@ use tracing::info;
...
         match &node.kind {
-            descriptor::CoreNodeKind::Custom(node) => match node.source.as_str() {
+            descriptor::CoreNodeKind::Custom(custom) => match custom.source.as_str() {
                 SHELL_SOURCE => (),
                 source => {
                     if source_is_url(source) {
                         info!("{source} is a URL."); // TODO: Implement url check.
+                    } else if let Some(remote_daemon_id) = remote_daemon_id {
+                        if remote_daemon_id.contains(&node.deploy.machine.as_str())
+                            || remote_daemon_id.contains(&String::default().as_str())
```
[the remark this answers is not in the document]
Comment (XxChang) [reply libraries/core/src/descriptor/validate.rs]: Nice suggestion, thanks!

**Review by phil-opp — APPROVED** [review]
Thanks!

**Pull request merged**